webcompat/static/js/lib/bugform.js

```diff
@@ -500,40 +506,95 @@ function BugForm() {
+      .then(this.addImageURL)
+      .then(dfd.resolve, dfd.reject);
 
-    this.uploadImage(
-      // grab the data URI from background-image property, since it's already
-      // in the DOM: 'url("data:image/....")'
-      this.getDataURIFromPreviewEl(),
-      _.bind(function(response) {
-        this.addImageURL(
-          response,
-          _.bind(function() {
-            var formEl = this.form.get(0);
-            // Calling submit() manually on the form won't contain details
-            // about which <button> was clicked (since one wasn't clicked).
-            // So we create a hidden <input> to pass along in the form data.
-            var hiddenEl = document.createElement("input");
-            hiddenEl.type = "hidden";
-            hiddenEl.name = "submit-type";
-            hiddenEl.value = this.clickedButton;
-            formEl.appendChild(hiddenEl);
-            formEl.submit();
-          }, this)
-        );
-      }, this)
-    );
+    return dfd.promise();
+  };
+
+  /*
+     Upload the image before form submission so we can
+     put an image link in the bug description.
+  */
+  this.uploadImage = function(dataURI) {
+    var dfd = $.Deferred();
+    this.disableSubmits();
+    this.uploadLoader.addClass("is-active");
+
+    $(".js-remove-upload").addClass("is-hidden");
+
+    var formdata = new FormData();
+    formdata.append("image", dataURI);
+
+    $.ajax({
+      contentType: false,
+      processData: false,
+      data: formdata,
+      method: "POST",
+      url: "/upload/"
+    }).then(dfd.resolve, dfd.reject);
+
+    return dfd.promise();
+  };
+
+  /*
+     React to server-side errors related to images by showing a flash
+     message to the user, and clearing out the bad image and preview.
+
+     If we're here, the attempted form submission failed.
+  */
+  this.handleUploadError = function(response) {
+    this.uploadLoader.removeClass("is-active");
+
+    if (response && response.status === 415) {
+      wcEvents.trigger("flash:error", {
+        message: this.inputs.image.helpText,
+        timeout: 5000
+      });
+    }
+
+    if (response && response.status === 413) {
+      wcEvents.trigger("flash:error", {
+        message:
+          "The image is too big! Please choose something smaller than 4MB.",
+        timeout: 5000
+      });
+    }
+
+    this.hideLoadingIndicator();
+    this.removeUploadPreview();
+  };
+
+  this.submitForm = function() {
+    var dfd = $.Deferred();
+    var formEl = this.form.get(0);
+    // Calling submit() manually on the form won't contain details
+    // about which <button> was clicked (since one wasn't clicked).
+    // So we create a hidden <input> to pass along in the form data.
+    var hiddenEl = document.createElement("input");
+    hiddenEl.type = "hidden";
+    hiddenEl.name = "submit-type";
+    hiddenEl.value = this.clickedButton;
+    formEl.appendChild(hiddenEl);
+    formEl.submit();
+    dfd.resolve();
+    return dfd.promise();
```

Comment:
Could we add the hidden input into the HTML of the form instead of generating it here and only update its value, when a submit button is clicked? It could contain the value `"anonymous"` initially.